eva/optimizer/column_deriver.py

```diff
@@ -0,0 +1,152 @@
+            prev_map.update(curr_map)
+            prev_merged.update(merged)
+        
+        return prev_map, prev_merged
+
+    def merge(self, map1: dict, map2: dict) -> dict:
+        fin_map = map1
+        for key,val in map2.items():
+            if key not in fin_map:
+                fin_map[key] = val
+        return fin_map
+
+    @scan_node_for_attributes.register(AbstractPlan)
+    def _for_abstract_plan(self, node:AbstractPlan, alias_prefix:str=None) -> dict:
+        return {},{}
+
+    @scan_node_for_attributes.register(ProjectPlan)
+    def _for_project_plan(self, node:ProjectPlan, alias_prefix:str=None) -> dict:
+        child_nodes = node.children
+        return self.scan_children(child_nodes)
+    
+    @scan_node_for_attributes.register(StoragePlan)
+    def _for_storage_plan(self, node:StoragePlan, alias_prefix:str=None):
+        """
+        Scans the children of the Storage Plan node. Also gets the columns from
+        video object.
+        Assumption: columns of video are placed after the columns of the childrens
+        """
+        node_children = node.predicate.children
+        column_name_to_id_mapping = {}
+        prev_map = {}
+
+        prev_map, merged_map = self.scan_children(node_children)
+            
+        video_columns_map = {}
+        attribute_id = 0
+        video_columns = node.video.columns
+        for column in video_columns:
+            col_name = alias_prefix+"."+column.name
+            video_columns_map[col_name] = attribute_id
+            attribute_id+=1
+        
+        column_name_to_id_mapping = self.merge(prev_map, video_columns_map)
+
+        return column_name_to_id_mapping, merged_map
+
+    @scan_node_for_attributes.register(SeqScanPlan)
+    def _for_seq_scan_plan(self, node:SeqScanPlan, alias_prefix:str=None) -> dict:
+        alias_prefix = node.alias.alias_name
+        return self.scan_children(node.children, alias_prefix)
+
+    @scan_node_for_attributes.register(LateralJoinPlan)
+    def _for_lateral_join_plan(self,node:LateralJoinPlan, alias_prefix:str=None) -> dict:
+        """
+        Scans the left child first and then the right child. Makes the assumption
+        that the left child is the first element of children and right child is the 
+        second element of children
+        """
+        left_child_map, _ = self.scan_node_for_attributes(node.children[0], alias_prefix)
+        right_child_map, _ = self.scan_node_for_attributes(node.children[1], alias_prefix)
+        col_update_map = left_child_map.copy()
+        merged_id_map = self.merge(left_child_map, right_child_map)
+        
+        #update column numbers
+        
+        attribute_id_counter = len(col_update_map)
+
+        for key,_ in right_child_map.items():
+            if key not in col_update_map:
+                col_update_map[key] = attribute_id_counter
+                attribute_id_counter+=1
+        
+        return merged_id_map, col_update_map
```

Comment:
Is the second output ever used?

Reply:
yes its used in the project plan on line 45 in ColumnMapper